Online status indicator

Parveshdhull opened the pull request:
fixes #12547

### Summary
PR implements Online Status Indicator
<div align="center">
<img src="https://user-images.githubusercontent.com/17097240/137238899-27f531a4-f463-4029-b526-8c906cb7d02f.png" alt="Screenshot" height="400"/>
<img src="https://user-images.githubusercontent.com/17097240/137976829-6e2dde90-bbc4-424c-b52a-9fdbc86afcdc.gif" alt="Screenshot" height="400"/>
<img src="https://user-images.githubusercontent.com/17097240/137238905-4bca1567-61f6-40c0-9e2e-d6456f6c4e23.png" alt="Screenshot" height="400"/>
</div>

PR follows these design guidelines: Designs [[1st]](https://www.figma.com/file/TNCyHKtR3sx5EL6YznFWUa4O/Profile?node-id=7253%3A27195) [[2nd]](https://www.figma.com/file/IPpvkpDWabBKJTeo6bFop0/Kuba%E2%8E%9CDesktop?node-id=77%3A19108)
#### status-go PR:
https://github.com/status-im/status-go/pull/2408

#### Code
There were some residue code in [styles.cljs](https://github.com/status-im/status-react/pull/12706/files#diff-f54bb761ef81cb3f9f1af35f7080f4edc667f1e88fb816755a5d14d7d442992d) file from 2016 implementation of online status indicator. Currently, it is not used anywhere, so removed to avoid confusion between styles.

#### Other features:
- Status updates syncing between devices
- Offline status updates
- Sort community members & contacts based on online status and name

#### Known Issues:
- Gap between each ping is 5 minutes. So it has its own side effects for ex. the user seems online a little longer, etc.
- Offline capability works great with the single device it will broadcast its current status whenever the device becomes online, even after app restart. 
But there is one rare case where it will not work:
i) User is using multiple devices
ii) Another device is online and broadcasting different status than saved by the main device during the offline state. (with larger clock value than saved clock)
iii) Once the main device gets online it receives broadcast from another device before broadcasting saved status. 

Thanks

rasom commented on src/status_im/status_updates/core.cljs:
```diff
@@ -0,0 +1,103 @@
+    (merge {:db (-> db
+                    (update-in [:status-updates] merge status-updates)
+                    (update-in [:multiaccount :current-user-status] merge current-user-status))}
+           (if (nil? dispatch) {} {:dispatch dispatch}))))
```
could be just `(when dispatch {:dispatch dispatch})`. It is shorter although still dispatch dispatch dispatch :D

rasom commented on src/status_im/status_updates/core.cljs:
```diff
@@ -0,0 +1,103 @@
+  {:events [:status-updates/visibility-status-option-pressed]}
+  [_ status-type]
+  (let [events-to-dispatch-later (atom [])]
+    (swap! events-to-dispatch-later conj {:ms 10 :dispatch [:status-updates/update-visibility-status status-type]})
```
When it is possible, please, keep lines' length shorter/equal 80 chars.

Parveshdhull replied:
Great Tip 🙏, now I enabled Emacs auto highlight for longer lines. Thanks

rasom commented on src/status_im/utils/datetime.cljs:
```diff
@@ -167,3 +169,6 @@
                  (s m)
                  (s 0)))))
+
+(defn to-mills [sec]
```
That's very subjective, but imo `to-ms` would look better.

Parveshdhull replied:
Yes, this looks much better 😅

flexsurfer commented on src/status_im/constants.cljs:
```diff
@@ -162,3 +162,9 @@
+(def ^:const visibility-status-automatic 1)
+(def ^:const visibility-status-dnd 2)
+(def ^:const visibility-status-always-online 3)
+(def ^:const visibility-status-inactive 4)
```
what about invisible ?

Parveshdhull replied:
Hi @flexsurfer, [originally](https://github.com/status-im/status-react/pull/12706), I had invisible. But that created conflicts with the desktop's [designs](https://www.figma.com/file/IPpvkpDWabBKJTeo6bFop0/Kuba%E2%8E%9CDesktop?node-id=77%3A19108).

flexsurfer replied:
I'm more interested in the UX, by default I would expect that my status is not propagated, is that the case ?

flexsurfer commented on src/status_im/data_store/settings.cljs:
```diff
@@ -50,4 +51,5 @@
       (update :custom-bootnodes-enabled? rpc->custom-bootnodes)
-      (update :currency keyword)))
+      (update :currency keyword)
+      (update :current-user-status status-updates/<-rpc)))
```
we have status tab. so probably we should use something like `user-online-status` or smth

Parveshdhull replied:
`user-online-status` will also create some confusion because there is also a status type as online. I will use `visibility-status-updates` instead of plain `status-updates` 👍

du82 replied:
> we have status tab. so probably we should use something like `user-online-status` or smth

I think we should rename the Status tab to "Moments" because it's more in line with WeChat, and it's less confusing to have a Status tab in the Status app.

flexsurfer commented on src/status_im/db.cljs:
```diff
@@ -32,6 +32,7 @@
              :tooltips                           {}
              :dimensions/window                  (dimensions/window)
              :registry                           {}
+             :status-updates                     {}
```
this might be confused with statuses (timeline)

status-im-auto commented:
## 11% of end-end tests have passed
```
Total executed tests: 19
Failed tests: 17
Passed tests: 2
```
```
IDs of failed tests: 5782,6645,6283,5341,6633,5783,5315,6253,5390,6263,5381,6240,5394,5758,5502,5356,5346 
```
<h3>Failed tests (17)</h3><details><summary>Click to expand</summary><br/><li><a href="https://ci.status.im/job/end-to-end-tests/job/status-app-prs-rerun/parambuild/?BRANCH_NAME=develop&APK_NAME=12716.apk&PR_ID=12716&TR_CASE_IDS=5782,6645,6283,5341,6633,5783,5315,6253,5390,6263,5381,6240,5394,5758,5502,5356,5346">Rerun tests</a></li><br/><table style="width: 100%"><colgroup><col span="1" style="width: 20%;"><col span="1" style="width: 80%;"></colgroup><tbody><tr></tr><tr><td><b>1. <a href="https://ethstatus.testrail.net/index.php?/tests/view/2332672">test_install_pack_and_send_sticker</a>, id: 5782 </b></td></tr><tr><td><p><blockquote><div>Device 2: **Sign in** (password:qwerty, keycard:False) </div><div>Device 2: *Find EditBox by accessibility id:* `password-input` </div></blockquote></p><code> Device 2: EditBox by accessibility id: `password-input` is not found on the screen; also Unexpected Alert is shown: 'Please shake the phone to report this error and restart the app. multiaccount is currently empty, which means something went wrong when t</code><br/><br/><ins>Device sessions</ins><p><ul><p>Device 1:<ul><li><a href="https://saucelabs.com/jobs/d1d35b1191b4468fb26bf6facecf1191?auth=cecd816db8ac3712c4af507a78b8a887">Steps, video, logs</a></li><li><a href="https://assets.saucelabs.com/jobs/d1d35b1191b4468fb26bf6facecf1191/0018screenshot.png?auth=cecd816db8ac3712c4af507a78b8a887">Failure screenshot</a></li></ul></p><p>Device 2:<ul><li><a href="https://saucelabs.com/jobs/efd985dfd6a04b39812f4158d3ba2bcc?auth=36bea4c01fefb89d4fe9fd741851de0f">Steps, video, logs</a></li><li><a href="https://assets.saucelabs.com/jobs/efd985dfd6a04b39812f4158d3ba2bcc/0019screenshot.png?auth=36bea4c01fefb89d4fe9fd741851de0f">Failure screenshot</a></li></ul></p></ul></p></td></tr><tr><td><b>2. <a href="https://ethstatus.testrail.net/index.php?/tests/view/2332680">test_restore_account_migrate_multiaccount_to_keycard</a>, id: 6645 </b></td></tr><tr><td><p><blockquote><div>Device 1: ========================================================================= </div><div>Device 1: *Find Button by accessibility id:* `back-button` </div></blockquote></p><code> Device 1: Button by accessibility id: `back-button` is not found on the screen; also Unexpected Alert is shown: 'Please shake the phone to report this error and restart the app. multiaccount is currently empty, which means something went wrong when tryin</code><br/><br/><ins>Device sessions</ins><p><ul><p>Device 1:<ul><li><a href="https://saucelabs.com/jobs/54a685a1b9884b46b075d6db127f24b1?auth=0aa361ce570afd0e55721ce9c3271871">Steps, video, logs</a></li><li><a href="https://assets.saucelabs.com/jobs/54a685a1b9884b46b075d6db127f24b1/0020screenshot.png?auth=0aa361ce570afd0e55721ce9c3271871">Failure screenshot</a></li></ul></p></ul></p></td></tr><tr><td><b>3. <a href="https://ethstatus.testrail.net/index.php?/tests/view/2332678">test_push_notification_1_1_chat_no_pn_activity_center</a>, id: 6283 </b></td></tr><tr><td><p><blockquote><div>Device 1: *Text is Welldocumented Gaseous Zigzagsalamander* </div><div>Device 1: *Find Button by accessibility id:* `notifications-settings-button` </div></blockquote></p><code> Device 1: Button by accessibility id: `notifications-settings-button` is not found on the screen</code><br/><br/><ins>Device sessions</ins><p><ul><p>Device 1:<ul><li><a href="https://saucelabs.com/jobs/188e503ecea2410bbc36f327fa07ba3b?auth=24d9092b9a0e5e472e4ca5bb36930028">Steps, video, logs</a></li><li><a href="https://assets.saucelabs.com/jobs/188e503ecea2410bbc36f327fa07ba3b/0012screenshot.png?auth=24d9092b9a0e5e472e4ca5bb36930028">Failure screenshot</a></li></ul></p><p>Device 2:<ul><li><a href="https://saucelabs.com/jobs/3d183720f1924e96b099555913b903dd?auth=6d4e1440d42cfaa1a6f1e0f90ab76499">Steps, video, logs</a></li><li><a href="https://assets.saucelabs.com/jobs/3d183720f1924e96b099555913b903dd/0011screenshot.png?auth=6d4e1440d42cfaa1a6f1e0f90ab76499">Failure screenshot</a></li></ul></p></ul></p></td></tr><tr><td><b>4. <a href="https://ethstatus.testrail.net/index.php?/tests/view/2332664">test_manage_assets</a>, id: 5341 </b></td></tr><tr><td><p><blockquote><div>Device 1: **Sign in** (password:qwerty, keycard:False) </div><div>Device 1: *Find EditBox by accessibility id:* `password-input` </div></blockquote></p><code> Device 1: EditBox by accessibility id: `password-input` is not found on the screen; also Unexpected Alert is shown: 'Please shake the phone to report this error and restart the app. multiaccount is currently empty, which means something went wrong when t</code><br/><br/><ins>Device sessions</ins><p><ul><p>Device 1:<ul><li><a href="https://saucelabs.com/jobs/c581f217e21a448baf434f8e1ce3e9c5?auth=b72097eac313d0b404f4880b9a487434">Steps, video, logs</a></li><li><a href="https://assets.saucelabs.com/jobs/c581f217e21a448baf434f8e1ce3e9c5/0021screenshot.png?auth=b72097eac313d0b404f4880b9a487434">Failure screenshot</a></li></ul></p></ul></p></td></tr><tr><td><b>5. <a href="https://ethstatus.testrail.net/index.php?/tests/view/2332679">test_browser_managing_bookmarks</a>, id: 6633 </b></td></tr><tr><td><p><blockquote><div>Device 1: **Sign in** (password:qwerty, keycard:False) </div><div>Device 1: *Find EditBox by accessibility id:* `password-input` </div></blockquote></p><code> Device 1: EditBox by accessibility id: `password-input` is not found on the screen; also Unexpected Alert is shown: 'Please shake the phone to report this error and restart the app. multiaccount is currently empty, which means something went wrong when t</code><br/><br/><ins>Device sessions</ins><p><ul><p>Device 1:<ul><li><a href="https://saucelabs.com/jobs/bd708392b07c409aaf09001e04c9a666?auth=28eab17ba668cdae829cceef0bc7ff03">Steps, video, logs</a></li><li><a href="https://assets.saucelabs.com/jobs/bd708392b07c409aaf09001e04c9a666/0034screenshot.png?auth=28eab17ba668cdae829cceef0bc7ff03">Failure screenshot</a></li></ul></p></ul></p></td></tr><tr><td><b>6. <a href="https://ethstatus.testrail.net/index.php?/tests/view/2332673">test_can_use_purchased_stickers_on_recovered_account</a>, id: 5783 </b></td></tr><tr><td><p><blockquote><div>Device 1: **Sign in** (password:qwerty, keycard:False) </div><div>Device 1: *Find EditBox by accessibility id:* `password-input` </div></blockquote></p><code> Device 1: EditBox by accessibility id: `password-input` is not found on the screen; also Unexpected Alert is shown: 'Please shake the phone to report this error and restart the app. multiaccount is currently empty, which means something went wrong when t</code><br/><br/><ins>Device sessions</ins><p><ul><p>Device 1:<ul><li><a href="https://saucelabs.com/jobs/484a09fcbafd4f7aa6ae17bd8418f4fc?auth=c0aee4c24fdbc54ce19e9528709c7d2a">Steps, video, logs</a></li><li><a href="https://assets.saucelabs.com/jobs/484a09fcbafd4f7aa6ae17bd8418f4fc/0022screenshot.png?auth=c0aee4c24fdbc54ce19e9528709c7d2a">Failure screenshot</a></li></ul></p></ul></p></td></tr><tr><td><b>7. <a href="https://ethstatus.testrail.net/index.php?/tests/view/2332662">test_send_non_english_message_to_newly_added_contact_on_different_networks</a>, id: 5315 </b></td></tr><tr><td><p><blockquote><div>Device 1: **Sign in** (password:qwerty, keycard:False) </div><div>Device 1: *Find EditBox by accessibility id:* `password-input` </div></blockquote></p><code> Device 1: EditBox by accessibility id: `password-input` is not found on the screen; also Unexpected Alert is shown: 'Please shake the phone to report this error and restart the app. multiaccount is currently empty, which means something went wrong when t</code><br/><br/><ins>Device sessions</ins><p><ul><p>Device 1:<ul><li><a href="https://saucelabs.com/jobs/c8be88f4a1be4f0a9e958c875c4ffa84?auth=169181f83219537ee404079ebed9df04">Steps, video, logs</a></li><li><a href="https://assets.saucelabs.com/jobs/c8be88f4a1be4f0a9e958c875c4ffa84/0018screenshot.png?auth=169181f83219537ee404079ebed9df04">Failure screenshot</a></li></ul></p><p>Device 2:<ul><li><a href="https://saucelabs.com/jobs/ae6b467e0c2442bbbf16da07a168fb8b?auth=582f48edc1892428b58945b0cbe1242b">Steps, video, logs</a></li><li><a href="https://assets.saucelabs.com/jobs/ae6b467e0c2442bbbf16da07a168fb8b/0011screenshot.png?auth=582f48edc1892428b58945b0cbe1242b">Failure screenshot</a></li></ul></p></ul></p></td></tr><tr><td><b>8. <a href="https://ethstatus.testrail.net/index.php?/tests/view/2332676">test_send_eth_in_1_1_chat_transaction_push</a>, id: 6253 </b></td></tr><tr><td><p><blockquote><div>Device 1: *Find ProfileButton by xpath:* `//*[contains(@content-desc,'5 out of 5')]` </div><div>Device 1: *Find Button by accessibility id:* `notifications-settings-button` </div></blockquote></p><code> Device 1: Button by accessibility id: `notifications-settings-button` is not found on the screen</code><br/><br/><ins>Device sessions</ins><p><ul><p>Device 1:<ul><li><a href="https://saucelabs.com/jobs/cc31ffe04bf04c66b824f33cba9c8dd1?auth=efa62e92b28fa47a256d8d49b0810d84">Steps, video, logs</a></li><li><a href="https://assets.saucelabs.com/jobs/cc31ffe04bf04c66b824f33cba9c8dd1/0016screenshot.png?auth=efa62e92b28fa47a256d8d49b0810d84">Failure screenshot</a></li></ul></p><p>Device 2:<ul><li><a href="https://saucelabs.com/jobs/ee9ef38e38104e9e9ecf4cac333fec31?auth=6acb1c3a99d3522be7b02e44f86ccaec">Steps, video, logs</a></li><li><a href="https://assets.saucelabs.com/jobs/ee9ef38e38104e9e9ecf4cac333fec31/0011screenshot.png?auth=6acb1c3a99d3522be7b02e44f86ccaec">Failure screenshot</a></li></ul></p></ul></p></td></tr><tr><td><b>9. <a href="https://ethstatus.testrail.net/index.php?/tests/view/2332668">test_delete_close_all_tabs</a>, id: 5390 </b></td></tr><tr><td><p><blockquote><div>Device 1: **Sign in** (password:qwerty, keycard:False) </div><div>Device 1: *Find EditBox by accessibility id:* `password-input` </div></blockquote></p><code> Device 1: EditBox by accessibility id: `password-input` is not found on the screen; also Unexpected Alert is shown: 'Please shake the phone to report this error and restart the app. multiaccount is currently empty, which means something went wrong when t</code><br/><br/><ins>Device sessions</ins><p><ul><p>Device 1:<ul><li><a href="https://saucelabs.com/jobs/c70a81dd4a424b55b80155e414fb7d5b?auth=a7891cfc363f13229e825ddc27a9bd3d">Steps, video, logs</a></li><li><a href="https://assets.saucelabs.com/jobs/c70a81dd4a424b55b80155e414fb7d5b/0031screenshot.png?auth=a7891cfc363f13229e825ddc27a9bd3d">Failure screenshot</a></li></ul></p></ul></p></td></tr><tr><td><b>10. <a href="https://ethstatus.testrail.net/index.php?/tests/view/2332677">test_request_and_receive_stt_in_1_1_chat_offline</a>, id: 6263 </b></td></tr><tr><td><p><blockquote><div>Device 2: **Sign in** (password:qwerty, keycard:False) </div><div>Device 2: *Find EditBox by accessibility id:* `password-input` </div></blockquote></p><code> Device 2: EditBox by accessibility id: `password-input` is not found on the screen; also Unexpected Alert is shown: 'Please shake the phone to report this error and restart the app. multiaccount is currently empty, which means something went wrong when t</code><br/><br/><ins>Device sessions</ins><p><ul><p>Device 1:<ul><li><a href="https://saucelabs.com/jobs/520eb3497de643c9ab090514fa1ab4d6?auth=4c33430c7e17302d9121f3282800cac8">Steps, video, logs</a></li><li><a href="https://assets.saucelabs.com/jobs/520eb3497de643c9ab090514fa1ab4d6/0031screenshot.png?auth=4c33430c7e17302d9121f3282800cac8">Failure screenshot</a></li></ul></p><p>Device 2:<ul><li><a href="https://saucelabs.com/jobs/af9bbd20c8274f34bbf4bf6120054005?auth=2655e98d40ebbc131c39552e3f98f747">Steps, video, logs</a></li><li><a href="https://assets.saucelabs.com/jobs/af9bbd20c8274f34bbf4bf6120054005/0029screenshot.png?auth=2655e98d40ebbc131c39552e3f98f747">Failure screenshot</a></li></ul></p></ul></p></td></tr><tr><td><b>11. <a href="https://ethstatus.testrail.net/index.php?/tests/view/2332667">test_user_can_see_collectibles_on_rinkeby_after_account_recovering</a>, id: 5381 </b></td></tr><tr><td><p><blockquote><div>Device 1: **Sign in** (password:qwerty, keycard:False) </div><div>Device 1: *Find EditBox by accessibility id:* `password-input` </div></blockquote></p><code> Device 1: EditBox by accessibility id: `password-input` is not found on the screen; also Unexpected Alert is shown: 'Please shake the phone to report this error and restart the app. multiaccount is currently empty, which means something went wrong when t</code><br/><br/><ins>Device sessions</ins><p><ul><p>Device 1:<ul><li><a href="https://saucelabs.com/jobs/f527839af6fb414ca21a9a7f8d6441ef?auth=4d12fadf677eb191cf15e722f6041c0b">Steps, video, logs</a></li><li><a href="https://assets.saucelabs.com/jobs/f527839af6fb414ca21a9a7f8d6441ef/0022screenshot.png?auth=4d12fadf677eb191cf15e722f6041c0b">Failure screenshot</a></li></ul></p></ul></p></td></tr><tr><td><b>12. <a href="https://ethstatus.testrail.net/index.php?/tests/view/2332674">test_restore_account_from_mnemonic_to_keycard</a>, id: 6240 </b></td></tr><tr><td><p><blockquote><div>Device 1: ========================================================================= </div><div>Device 1: **Sign in** (password:qwerty, keycard:True) </div></blockquote></p><code> Device 1: SilentButton by accessibility id:`numpad-button-1` is not found on the screen; also Unexpected Alert is shown: 'Please shake the phone to report this error and restart the app. multiaccount is currently empty, which means something went wrong w</code><br/><br/><ins>Device sessions</ins><p><ul><p>Device 1:<ul><li><a href="https://saucelabs.com/jobs/4ea48f597cf549f6ac2de90c64f56c12?auth=650792e6683e7681beb98178dd57c3b5">Steps, video, logs</a></li><li><a href="https://assets.saucelabs.com/jobs/4ea48f597cf549f6ac2de90c64f56c12/0039screenshot.png?auth=650792e6683e7681beb98178dd57c3b5">Failure screenshot</a></li></ul></p></ul></p></td></tr><tr><td><b>13. <a href="https://ethstatus.testrail.net/index.php?/tests/view/2332669">test_account_recovery_with_uppercase_whitespaces_seed_phrase_special_char_passw_logcat</a>, id: 5394 </b></td></tr><tr><td><p><blockquote><div>Device 1: **Sign in** (password: !"#$Á%Ö&'()*+Í, -./:ä;<=>?@[\]^_`{|}~ , keycard:False) </div><div>Device 1: *Find EditBox by accessibility id:* `password-input` </div></blockquote></p><code> Device 1: EditBox by accessibility id: `password-input` is not found on the screen; also Unexpected Alert is shown: 'Please shake the phone to report this error and restart the app. multiaccount is currently empty, which means something went wrong when t</code><br/><br/><ins>Device sessions</ins><p><ul><p>Device 1:<ul><li><a href="https://saucelabs.com/jobs/87ddfffaabbf41a480950e8f446d3090?auth=30f50e6a2521e1481a20a313d9cce7d2">Steps, video, logs</a></li><li><a href="https://assets.saucelabs.com/jobs/87ddfffaabbf41a480950e8f446d3090/0025screenshot.png?auth=30f50e6a2521e1481a20a313d9cce7d2">Failure screenshot</a></li></ul></p></ul></p></td></tr><tr><td><b>14. <a href="https://ethstatus.testrail.net/index.php?/tests/view/2332671">test_keycard_can_recover_keycard_account_card_pairing</a>, id: 5758 </b></td></tr><tr><td><p><blockquote><div>Device 1: ========================================================================= </div><div>Device 1: **Sign in** (password:qwerty, keycard:True) </div></blockquote></p><code> Device 1: SilentButton by accessibility id:`numpad-button-1` is not found on the screen; also Unexpected Alert is shown: 'Please shake the phone to report this error and restart the app. multiaccount is currently empty, which means something went wrong w</code><br/><br/><ins>Device sessions</ins><p><ul><p>Device 1:<ul><li><a href="https://saucelabs.com/jobs/3ea4a636b4c7457e84d3924335ba931b?auth=9701f205c833756763cb3f715e53983f">Steps, video, logs</a></li><li><a href="https://assets.saucelabs.com/jobs/3ea4a636b4c7457e84d3924335ba931b/0027screenshot.png?auth=9701f205c833756763cb3f715e53983f">Failure screenshot</a></li></ul></p></ul></p></td></tr><tr><td><b>15. <a href="https://ethstatus.testrail.net/index.php?/tests/view/2332670">test_can_add_existing_ens_on_mainnet</a>, id: 5502 </b></td></tr><tr><td><p><blockquote><div>Device 1: **Sign in** (password:qwerty, keycard:False) </div><div>Device 1: *Find EditBox by accessibility id:* `password-input` </div></blockquote></p><code> Device 1: EditBox by accessibility id: `password-input` is not found on the screen; also Unexpected Alert is shown: 'Please shake the phone to report this error and restart the app. multiaccount is currently empty, which means something went wrong when t</code><br/><br/><ins>Device sessions</ins><p><ul><p>Device 1:<ul><li><a href="https://saucelabs.com/jobs/07212f5a6b894a328b7e1285969e7b50?auth=86747ff23c78b4264a41d60a95a6e955">Steps, video, logs</a></li><li><a href="https://assets.saucelabs.com/jobs/07212f5a6b894a328b7e1285969e7b50/0022screenshot.png?auth=86747ff23c78b4264a41d60a95a6e955">Failure screenshot</a></li></ul></p></ul></p></td></tr><tr><td><b>16. <a href="https://ethstatus.testrail.net/index.php?/tests/view/2332666">test_switch_users_special_char_password_and_add_new_account_logcat</a>, id: 5356 </b></td></tr><tr><td><p><blockquote><div>Device 1: ========================================================================= </div><div>Device 1: *Find MultiAccountOnLoginButton by xpath:* `(//*[@content-desc='chat-icon'])[1]/..` </div></blockquote></p><code> Device 1: MultiAccountOnLoginButton by xpath: `(//*[@content-desc='chat-icon'])[1]/..` is not found on the screen; also Unexpected Alert is shown: 'Please shake the phone to report this error and restart the app. multiaccount is currently empty, which me</code><br/><br/><ins>Device sessions</ins><p><ul><p>Device 1:<ul><li><a href="https://saucelabs.com/jobs/4fdd8c3378d049559272622c8b3a318d?auth=608a6d8cd2fbe63236b9cb71587c47af">Steps, video, logs</a></li><li><a href="https://assets.saucelabs.com/jobs/4fdd8c3378d049559272622c8b3a318d/0017screenshot.png?auth=608a6d8cd2fbe63236b9cb71587c47af">Failure screenshot</a></li></ul></p></ul></p></td></tr><tr><td><b>17. <a href="https://ethstatus.testrail.net/index.php?/tests/view/2332665">test_collectible_from_wallet</a>, id: 5346 </b></td></tr><tr><td><p><blockquote><div>Device 1: **Sign in** (password:qwerty, keycard:False) </div><div>Device 1: *Find EditBox by accessibility id:* `password-input` </div></blockquote></p><code> Device 1: EditBox by accessibility id: `password-input` is not found on the screen; also Unexpected Alert is shown: 'Please shake the phone to report this error and restart the app. multiaccount is currently empty, which means something went wrong when t</code><br/><br/><ins>Device sessions</ins><p><ul><p>Device 1:<ul><li><a href="https://saucelabs.com/jobs/2f4640da1c354094beb6790d93d77736?auth=558166da4ca83518bed041f3b85999b9">Steps, video, logs</a></li><li><a href="https://assets.saucelabs.com/jobs/2f4640da1c354094beb6790d93d77736/0029screenshot.png?auth=558166da4ca83518bed041f3b85999b9">Failure screenshot</a></li></ul></p></ul></p></td></tr></tbody></table></details><h3>Passed tests (2)</h3><details><summary>Click to expand</summary><br/><br/><table style="width: 100%"><colgroup><col span="1" style="width: 20%;"><col span="1" style="width: 80%;"></colgroup><tbody><tr></tr><tr><td><b>1. <a href="https://ethstatus.testrail.net/index.php?/tests/view/2332663">test_long_press_to_delete_chat</a>, id: 5319 </b></td></tr><tr><td><ins>Device sessions</ins><p><ul><p>Device 1:<ul><li><a href="https://saucelabs.com/jobs/3a8aea7a04d9404681657c43f54f0b67?auth=3f61b98425abbb36213e826171406e98">Steps, video, logs</a></li></ul></p></ul></p></td></tr><tr><td><b>2. <a href="https://ethstatus.testrail.net/index.php?/tests/view/2332675">test_keycard_request_stt_from_daap</a>, id: 6249 </b></td></tr><tr><td><ins>Device sessions</ins><p><ul><p>Device 1:<ul><li><a href="https://saucelabs.com/jobs/6721af44a2b544c7a5c3f7a392c8f3b9?auth=d183b3179b6614b98e393cb49b7da9f6">Steps, video, logs</a></li></ul></p></ul></p></td></tr></tbody></table></details>

flexsurfer commented on src/status_im/visibility_status_popover/core.cljs:
```diff
@@ -0,0 +1,21 @@
+(ns status-im.visibility-status-popover.core
```
I'm not sure if it is worth it to use a separate overlay for this, probably better to just use absolute layout

Parveshdhull replied:
I tried with the absolute layout too, but I was unable to make it the same as [designs](https://www.figma.com/file/TNCyHKtR3sx5EL6YznFWUa4O/Profile?node-id=7253%3A27195). The design requires a full overlay which also closes this on touch.

flexsurfer replied:
ok, thanks

flexsurfer commented on src/status_im/data_store/settings.cljs:
```diff
@@ -50,4 +51,5 @@
       (update :custom-bootnodes-enabled? rpc->custom-bootnodes)
-      (update :currency keyword)))
+      (update :currency keyword)
+      (update :current-user-status visibility-status-updates/<-rpc)))
```
:current-user-**visibility**-status ?

flexsurfer commented on src/status_im/ui/screens/profile/visibility_status/styles.cljs:
```diff
@@ -0,0 +1,70 @@
+(ns status-im.ui.screens.profile.visibility-status.styles
+  (:require [quo.design-system.colors :as colors]))
+
+(def color-error "#FA6565")
```
should we move this to color ns ?

flexsurfer commented on src/status_im/ui/screens/profile/visibility_status/styles.cljs:
```diff
@@ -0,0 +1,70 @@
+   :margin-left            16
+   :border-radius          16
+   :border-top-left-radius 4
+   :align-self             "flex-start"
```
better to use keywords :flex-start

flexsurfer commented on src/status_im/ui/screens/profile/visibility_status/styles.cljs:
```diff
@@ -0,0 +1,70 @@
+   :height           size
+   :border-radius    (/ size 2)
+   :border-width     1
+   :border-color     "#FFFFFF"})
```
colors/white

Parveshdhull replied:
Hi, can you please take a look at colors.cljs. It has the same value for white and white-persist and there is no pure white. That's why I hard-coded this value.

flexsurfer replied:
what do you mean pure white ?

Parveshdhull replied:
Pure White - #FFFFFF
current white is not actually white, it is changing based on theme

flexsurfer replied:
if you want white for both dark and light use white-persist  then

Parveshdhull replied:
Oh sorry, I was confused between theme and light theme. I will user colors/white, instead of hard coding 👍

Parveshdhull replied:
still it is little weird that white and white-persist are exactly same

flexsurfer commented on src/status_im/visibility_status_updates/core.cljs:
```diff
@@ -0,0 +1,168 @@
+                         visibility-status-update-loaded)]
+                    (assoc-in prev-map [:visibility-status-updates public-key]
+                              visibility-status-update)))
+                {:visibility-status-updates {}}
```
why dont just have {} here ? why do you need `:visibility-status-updates` key ?

flexsurfer commented on src/status_im/visibility_status_updates/core.cljs:
```diff
@@ -0,0 +1,168 @@
+  {:events [::visibility-status-updates-loaded]}
+  [{:keys [db]} visibility-status-updates-loaded]
+  (let [{:keys [visibility-status-updates]}
+        (reduce (fn [prev-map visibility-status-update-loaded]
```
usually prev-map named as `acc` - accumulator

flexsurfer commented on src/status_im/visibility_status_updates/core.cljs:
```diff
@@ -0,0 +1,168 @@
+(defn handle-my-visibility-status-updates
+  [prev-map my-current-status clock visibility-status-update]
+  (let [status-type (:status-type visibility-status-update)]
+    (when (or
```
dont you need to return prev-map for else case ?

flexsurfer commented on src/status_im/visibility_status_updates/core.cljs:
```diff
@@ -0,0 +1,168 @@
+           (when dispatch {:dispatch dispatch}))))
+
+(fx/defn initialize-visibility-status-updates [cofx]
+  (visibility-status-updates-store/fetch-visibility-status-updates-rpc
```
why not call this method directly ?

Parveshdhull replied:
I was trying to keep `login/core.cljs` clean and use the same pattern used earlier, for ex.  `(contact/initialize-contacts)`

Parveshdhull replied:
Now as I think, accessing directly also looks great 🤔, will change this too 👍 Thanks

flexsurfer commented on src/status_im/visibility_status_updates/core.cljs:
```diff
@@ -0,0 +1,168 @@
+  {:events [:visibility-status-updates/visibility-status-option-pressed]}
+  [{:keys [db]} status-type]
+  (let [peers-count              (:peers-count db)
+        events-to-dispatch-later (atom [])]
```
this doesn't look clojure idiomatic, whats the point of using atom here ?

flexsurfer commented on src/status_im/visibility_status_updates/core.cljs:
```diff
@@ -0,0 +1,168 @@
+(fx/defn countdown-for-online-user
+  {:events [:visibility-status-updates/countdown-for-online-user]}
+  [{:keys [db]} public-key clock ms]
+  {:dispatch-later
```
what is the point of having dispach-later inside dispatch?

flexsurfer commented on src/status_im/visibility_status_updates/core.cljs:
```diff
@@ -0,0 +1,168 @@
+           send-visibility-status-updates?
+           (= status-type constants/visibility-status-inactive))
+      {:dispatch-later
+       [{:ms 100
```
why dispatch-later and magic numbers 100 and 1000 ?

Parveshdhull replied:
I think the first dispatch-later can be changed into dispatch but the second one is stopping broadcasts once the current status is broadcast-ed. 1000 was just a safety net, can be changed into 100 or something I guess 🤔

Parveshdhull commented:
Hi @qoqobolo , Thank you very much for testing PR 🙏.
I fixed the above issues and also added another functionality.
`Sort community members & contacts based on online status and name`
Please let me know if I can improve anything else. Thanks

flexsurfer commented on src/status_im/data_store/settings.cljs:
```diff
@@ -50,4 +51,5 @@
       (update :custom-bootnodes-enabled? rpc->custom-bootnodes)
-      (update :currency keyword)))
+      (update :currency keyword)
+      (visibility-status-updates/<-rpc-settings)))
```
why is it different from others ? i mean why not `update`

Parveshdhull replied:
Hi, data from go is returned as

``` 
-current-user-status
	- statusType
	- clock
```

I can use update as `(update :current-user-status visibility-status-updates/<-rpc)`, but that will only change `statusType` into `status-type`
I also want to change key `current-user-status` into `current-user-visibility-status` and `<-rpc-settings` does both those things in one call

flexsurfer commented on src/status_im/ui/screens/contacts_list/views.cljs:
```diff
@@ -28,6 +30,14 @@
+(defn visibility-status-order [contact]
+  (visibility-status-utils/visibility-status-order (:public-key contact)))
+
+(defn sorted-contacts-list [contacts]
```
its better to create a new subscription, views must be dumb

flexsurfer commented on src/status_im/ui/screens/communities/members.cljs:
```diff
@@ -85,16 +87,27 @@
+(defn visibility-status-order [member]
+  (visibility-status-utils/visibility-status-order (get member 0)))
+
+(defn sorted-members-list [members]
```
better to move to subscriptions

flexsurfer commented on src/status_im/ui/screens/communities/members.cljs:
```diff
@@ -85,16 +87,27 @@
      [quo/separator {:style {:margin-vertical 8}}]]))
 
+(defn member-name [member]
+  (<sub [:contacts/contact-name-by-identity (get member 0)]))
```
this looks not correct, better to have one subscription

flexsurfer commented on src/status_im/visibility_status_updates/core.cljs:
```diff
@@ -0,0 +1,179 @@
+                      [:visibility-status-updates public-key]
+                      visibility-status-update)
+      (= real-status-type constants/visibility-status-automatic)
+      (assoc :dispatch-later
```
why not update ?

flexsurfer replied:
and still not sure why dispatch-later is used here ?

Parveshdhull replied:
1. assoc is used because dispatch-later is initially nil and update will through error
2. User is online for 5 minutes, dispatch-later is used to make him/her offline after 5 minutes

flexsurfer replied:
1. clojure doesn't through errors ;) it will work fine even if dispatch-later  is nil
2. hm now its even more confusing, could you please elaborate

Parveshdhull replied:
I also tried just checking if the online status is 5 minutes old directly on view, that works great, but the view doesn't re-render after 5 minutes, So I need to use dispatch-later to expire that status.

Parveshdhull replied:
Sure,
There are 3 status types
1. **Inactive** - User will always displayed as offline, and no ping is received - _Anonymous_
2. **Always Online** - User will always displayed as online, and online ping is received every 5 minutes (when sender is online) - _Pseudo Anonymous_
3. **Automatic** - User will only be shown online if there is a online ping in last 5 minutes, otherwise offline

So for `automatic`, to make sure the user is online or not, I have to make sure if ping received with the clock is in last 5 minutes, and I have to convert that status into offline, after whatever time is left between `now` and `received clock`

Parveshdhull replied:
Yes it is confusing :D, you can find more details in [specs](https://github.com/status-im/status-go/pull/2408)

flexsurfer replied:
does that mean that for every process-visibility-status-update you 'll create a new dispatch later ? but you probably need to remove old dispatches ?

Parveshdhull replied:
Actually, another user is broadcasting at the rate of 5 minutes and this is dispatch-later for 5.05 minutes. So there will not be much issue of having many pending dispatch-laters because this one will expire whenever new will receive.

And also this dispatch checks clock of current status update before converting it into offline. If a new status is received, then it will not do anything.

So in a way there is only one dispatch-later per user, that keeps getting renewed every 5 minutes if another user is online and destroys after one run if he/she is not

Also dispatch-later will not be registered for inactive users, always online, and expired automatic. So it will only be registered for a fraction of user's, who are online at that moment.

flexsurfer commented on src/status_im/visibility_status_updates/core.cljs:
```diff
@@ -0,0 +1,179 @@
+           (> peers-count 0)
+           send-visibility-status-updates?
+           (= status-type constants/visibility-status-inactive))
+      {:dispatch [:visibility-status-updates/update-visibility-status status-type]
```
can't we use update-visibility-status handler here instead of dispatch?

flexsurfer commented on src/status_im/subs.cljs:
```diff
@@ -294,6 +295,16 @@
+ (fn [communities [_ id]]
+   (->> (get-in communities [id :members])
+        (sort-by (fn [member]
+                   @(re-frame/subscribe
```
unfortunately, this won't work properly, we have to create a new subscription :contacts/contact-names-by-identitis

flexsurfer commented on src/status_im/subs.cljs:
```diff
@@ -296,13 +296,22 @@
+   (let [members (re-frame/subscribe
+                  [:communities/community-members community-id])
+         names   (re-frame/subscribe
+                  [:contacts/contact-names-by-identities (keys @members)])]
```
:communities/community-members should be inside contact-names-by-identities

flexsurfer replied:
or you don't need contact-names-by-identities then you could move its code here

Parveshdhull replied:
Yes, better to remove `contact-names-by-identities`, because this function also requires members

flexsurfer commented on src/status_im/subs.cljs:
```diff
@@ -292,6 +295,37 @@
+         multiaccount (re-frame/subscribe [:multiaccount])
+         members      (re-frame/subscribe
+                       [:communities/community-members community-id])
+         names        (atom
```
why this reduce here with atom but not in the subscription function?

Parveshdhull committed:
Online Status Indicator
